src/finchlite/supertensor/interpreter.py

```diff
@@ -0,0 +1,216 @@
+import operator
+
+import numpy as np
+import copy
+
+from ..algebra import overwrite, promote_max, promote_min
+from typing import List, Tuple, Set, FrozenSet, Dict
+from itertools import chain, combinations
+
+from finchlite import finch_einsum as ein
+from finchlite import symbolic as sym
+from . import supertensor as stns
+
+class SuperTensorEinsumInterpreter:
+    def __init__(self, xp=None, bindings=None):
+        if bindings is None:
+            bindings = {}
+        if xp is None:
+            xp = np
+        self.bindings = bindings
+        self.xp = xp
+
+    def __call__(self, node):
+        match node:
+            case ein.Plan(bodies):
+                res = None
+                for body in bodies:
+                    res = self(body)
+                return res
+            case ein.Produces(args):
+                return tuple(self(arg) for arg in args)
+            case ein.Einsum(op, ein.Alias(output_name), output_idxs, arg):
+                accesses = SuperTensorEinsumInterpreter._collect_accesses(arg)
+
+                # Group the indices which appear in exactly the same set of tensors.
+                output_idxs = [idx.name for idx in output_idxs]
+
+                input_idx_list = []
+                for access in accesses:
+                    tns_name = access.tns.name
+                    idxs = [idx.name for idx in access.idxs]
+                    input_idx_list.append((tns_name, idxs))
+                
+                idx_groups = SuperTensorEinsumInterpreter._group_indices(output_name, output_idxs, input_idx_list)
+
+                # Assign a new index name to each group of original indices.
+                new_idxs = {}
+                for k, (tensor_set, _) in enumerate(idx_groups):
+                    new_idxs[tensor_set] = f"i{k}"
+
+
+                # Compute the corrected SuperTensor representation for each tensor.
+                inputs = []
+                for access in accesses:
+                    tns_name = access.tns.name
+                    supertensor = self.bindings[access.tns.name]
+                    idxs = [idx.name for idx in access.idxs]
+                    inputs.append((tns_name, supertensor, idxs))
+
+                corrected_bindings = {}
+                corrected_idx_lists = {}
+                for tns_name, supertensor, input_idx_list in inputs:
+                    new_idx_list = []
+                    mode_map = []
```

Comment:
we can construct a dictionary globally, which maps idx -> newidx, which I think would be helpful here.